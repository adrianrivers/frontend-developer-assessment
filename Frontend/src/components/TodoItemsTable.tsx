import React from 'react'
import { Table, Thead, Tbody, Tr, Th, Td, Box, Text, HStack } from '@chakra-ui/react'
import { TodoItem } from './TodoList'
import TodoDescription from './TodoDescription'
import TodoComplete from './TodoComplete'
interface TodoItemsTableProps {
  todoItems: TodoItem[]
  handleMarkAsCompleted: (id: number, description: string) => void
}

export default function TodoItemsTable({ todoItems, handleMarkAsCompleted }: TodoItemsTableProps) {
  const table = () => {
    return (
      <Table variant="simple">
        <Thead>
          <Tr>
            <Th>Description</Th>
            <Th isNumeric>Action</Th>
          </Tr>
        </Thead>
        <Tbody>
          {todoItems.reverse().map((todo: TodoItem) => {
            const { id, description, isCompleted } = todo
            return (
              <Tr key={id}>
                <Td>
                  <TodoDescription description={description} isCompleted={isCompleted} />
                </Td>
                <Td isNumeric marginRight={8}>
                  <TodoComplete
                    handleMarkAsCompleted={handleMarkAsCompleted}
                    isCompleted={isCompleted}
                    id={id}
                    description={description}
                  />
                </Td>
              </Tr>
            )
          })}
        </Tbody>
      </Table>
    )
  }

  return (
    <>
      <HStack paddingX={6} marginBottom={12} justifyContent="space-between">
        <Text fontWeight="bold">Showing {todoItems.length} Item(s) </Text>
      </HStack>
      <Box marginBottom={16}>
        {todoItems.length ? table() : <Text marginX={6}>You currently have no todo items in your todo list ðŸ™‚</Text>}
      </Box>
    </>
  )
}
