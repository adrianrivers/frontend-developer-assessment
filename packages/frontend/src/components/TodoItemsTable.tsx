import React, { useEffect, useState } from 'react'
import { Table, Thead, Tbody, Tr, Th, Td, Box, Text, HStack } from '@chakra-ui/react'
import TodoDescription from './TodoDescription'
import TodoComplete from './TodoComplete'
import useSWR from 'swr'
import { TODO_LIST_API } from '../constants'
import axios from 'axios'
import Confetti from './Confetti'
import { TodoItem } from '../types'

interface TableComponentProps {
  data: TodoItem[]
}

function ItemsTableBody({ data }: TableComponentProps): React.ReactElement {
  const tableRows = []

  for (const item of data) {
    const { id, description, isCompleted } = item

    tableRows.push(
      <Tr key={id}>
        <Td>
          <TodoDescription description={description} isCompleted={isCompleted} />
        </Td>
        <Td isNumeric marginRight={8}>
          <TodoComplete isCompleted={isCompleted} id={id} description={description} />
        </Td>
      </Tr>
    )
  }

  return <Tbody>{tableRows}</Tbody>
}

function ItemsTable({ data }: TableComponentProps): React.ReactElement {
  return (
    <>
      <HStack paddingX={6} marginBottom={12} justifyContent="space-between">
        <Text fontWeight="bold">Showing {data.length} Item(s)</Text>
      </HStack>
      <Table variant="simple">
        <Thead>
          <Tr>
            <Th>Description</Th>
            <Th isNumeric>Action</Th>
          </Tr>
        </Thead>
        <ItemsTableBody data={data} />
      </Table>
    </>
  )
}

export default function TodoItemsTable(): React.ReactElement {
  const [showConfetti, setShowConfetti] = useState(false)

  const fetcher = (url: string) => axios.get<TodoItem[]>(url).then((response) => response.data.reverse())
  const { data, error } = useSWR(TODO_LIST_API.DEV_URL, fetcher)

  const RenderContent = () => {
    if (data) {
      return <ItemsTable data={data} />
    } else if (!data) {
      return <Text marginX={6}>You currently have no todo items in your todo list ðŸ™‚</Text>
    } else if (error) {
      return <Text marginX={6}>Something went wrong fetching your todos ðŸ˜¢</Text>
    }
  }

  useEffect(() => {
    if (!data || data.length < 1) return

    const allTodoItemsCompleted = data.every((todoItem) => todoItem.isCompleted)

    if (allTodoItemsCompleted) {
      setShowConfetti(true)
    } else {
      setShowConfetti(false)
    }
  }, [data])

  return (
    <>
      {showConfetti && <Confetti />}
      <Box marginBottom={16}>{RenderContent()}</Box>
    </>
  )
}
