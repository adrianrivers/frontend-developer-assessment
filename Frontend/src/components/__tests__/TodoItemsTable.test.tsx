import { render, fireEvent, waitFor } from '@testing-library/react'
import React from 'react'
import TodoItemsTable from '../TodoItemsTable'
import { TodoItem } from '../../types'

const TodoItemsMock: TodoItem[] = [
  {
    id: 1,
    description: 'Buy milk',
    isCompleted: false,
  },
  {
    id: 2,
    description: 'Buy cheese',
    isCompleted: false,
  },
]

const handleMarkAsCompletedMock = jest.fn()

describe('TodoItemsTable', () => {
  it('should NOT render a table of todo items if none are passed to the component', () => {
    const { container } = render(<TodoItemsTable todoItems={[]} handleMarkAsCompleted={handleMarkAsCompletedMock} />)

    expect(container.querySelector('table')).toBeNull()
    expect(container).toMatchSnapshot()
  })

  it('should render a TodoItemsTable component', async () => {
    const { container, getByText, queryByText, getAllByText } = render(
      <TodoItemsTable todoItems={TodoItemsMock} handleMarkAsCompleted={handleMarkAsCompletedMock} />
    )
    const markAsCompletedButtons = getAllByText('Mark as completed') as HTMLButtonElement[]

    getByText(TodoItemsMock[0].description)
    expect(markAsCompletedButtons.length).toBe(2)

    fireEvent.click(markAsCompletedButtons[0])

    waitFor(() => {
      expect(queryByText(TodoItemsMock[0].description)).toBe(null)
      expect(getByText('Woohoo', { exact: false })).toBeInTheDocument()
    })

    expect(handleMarkAsCompletedMock).toHaveBeenCalledTimes(1)
    expect(handleMarkAsCompletedMock).toHaveBeenCalledWith(TodoItemsMock[0].id, TodoItemsMock[0].description)
    expect(container).toMatchSnapshot()
  })
})
