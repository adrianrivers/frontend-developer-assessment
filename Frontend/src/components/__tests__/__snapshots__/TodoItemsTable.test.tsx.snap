// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItemsTable should NOT render a table of todo items if none are passed to the component 1`] = `
<div>
  <div
    class="chakra-stack css-1eoh7ji"
  >
    <p
      class="chakra-text css-in3yi3"
    >
      Showing 
      0
       Item(s)
    </p>
  </div>
  <div
    class="css-1qm1lh"
  >
    <p
      class="chakra-text css-1xsyj6s"
    >
      You currently have no todo items in your todo list ðŸ™‚
    </p>
  </div>
</div>
`;

exports[`TodoItemsTable should render a TodoItemsTable component 1`] = `
<div>
  <div
    class="chakra-stack css-1eoh7ji"
  >
    <p
      class="chakra-text css-in3yi3"
    >
      Showing 
      2
       Item(s)
    </p>
  </div>
  <div
    class="css-1qm1lh"
  >
    <table
      class="chakra-table css-0"
      role="table"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <th
            class="css-0"
          >
            Description
          </th>
          <th
            class="css-0"
            data-is-numeric="true"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-0"
            role="gridcell"
          >
            <p
              class="chakra-text css-0"
            >
              Buy cheese
            </p>
          </td>
          <td
            class="css-12z0wuy"
            data-is-numeric="true"
            role="gridcell"
          >
            <button
              class="chakra-button css-16jemn4"
              type="button"
            >
              Mark as completed
            </button>
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-0"
            role="gridcell"
          >
            <p
              class="chakra-text css-0"
            >
              Buy milk
            </p>
          </td>
          <td
            class="css-12z0wuy"
            data-is-numeric="true"
            role="gridcell"
          >
            <button
              class="chakra-button css-16jemn4"
              type="button"
            >
              Mark as completed
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
